Sources/SwiftProtobuf/JSONScanner.swift

```diff
@@ -706,6 +398,311 @@ internal struct JSONScanner {
+    }
+    let c = source[index]
+    if c == asciiMinus { // -
+      source.formIndex(after: &index)
+      // character after '-' must be digit
+      let digit = source[index]
+      if digit < asciiZero || digit > asciiNine {
+        throw JSONDecodingError.malformedNumber
+      }
+      if let n = try parseBareUInt(source: source, index: &index, end: end) {
+        let limit: UInt64 = 0x8000000000000000 // -Int64.min
+        if n >= limit {
+          if n > limit {
+            // Too large negative number
+            throw JSONDecodingError.numberRange
+          } else {
+            return Int64.min // Special case for Int64.min
+          }
+        }
+        return -Int64(bitPattern: n)
+      } else {
+        return nil
+      }
+    } else if let n = try parseBareUInt(source: source, index: &index, end: end) {
+      if n > UInt64(bitPattern: Int64.max) {
+        throw JSONDecodingError.numberRange
+      }
+      return Int64(bitPattern: n)
+    } else {
+      return nil
+    }
+  }
+
+  // Identify a floating-point token in the upcoming UTF8 bytes.
+  //
+  // This implements the full grammar defined by the JSON RFC 7159.
+  // Note that Swift's string-to-number conversions are much more
+  // lenient, so this is necessary if we want to accurately reject
+  // malformed JSON numbers.
+  //
+  // This is used by nextDouble() and nextFloat() to parse double and
+  // floating-point values, including values that happen to be in quotes.
+  // It's also used by the slow path in parseBareSInt() and parseBareUInt()
+  // above to handle integer values that are written in float-point notation.
+  private func parseBareDouble(
+    source: UnsafeBufferPointer<UInt8>,
+    index: inout UnsafeBufferPointer<UInt8>.Index,
+    end: UnsafeBufferPointer<UInt8>.Index
+  ) throws -> Double? {
+    // RFC 7159 defines the grammar for JSON numbers as:
+    // number = [ minus ] int [ frac ] [ exp ]
+    let start = index
+    var c = source[index]
+    if c == asciiBackslash {
+      return nil
+    }
+
+    // Optional leading minus sign
+    if c == asciiMinus { // -
+      source.formIndex(after: &index)
+      if index == end {
+        index = start
+        throw JSONDecodingError.truncated
+      }
+      c = source[index]
+      if c == asciiBackslash {
+        return nil
+      }
+    } else if c == asciiUpperN { // Maybe NaN?
+      // Return nil, let the caller deal with it.
+      return nil
+    }
+
+    if c == asciiUpperI { // Maybe Infinity, Inf, -Infinity, or -Inf ?
+      // Return nil, let the caller deal with it.
+      return nil
+    }
+
+    // Integer part can be zero or a series of digits not starting with zero
+    // int = zero / (digit1-9 *DIGIT)
+    switch c {
+    case asciiZero:
+      // First digit can be zero only if not followed by a digit
+      source.formIndex(after: &index)
+      if index == end {
+        if let d = numberFormatter.utf8ToDouble(bytes: source, start: start, end: index) {
```

Comment:
Good catch!  It is always zero here.

I've also filled in some tests to verify that we do actually correctly parse 0 and -0.